libs/s25main/network/GameClient.h

```diff
@@ -129,6 +130,12 @@ class GameClient final :
 
     /// Lädt ein Replay und startet dementsprechend das Spiel
     bool StartReplay(const boost::filesystem::path& path);
+
+    /// When a non-empty vector is given then an AI battle with the given AIs is started
+    void SetAIBattlePlayers(std::vector<AI::Info>&& aiInfos);
```

Comment:
You are right, this approach will still create a copy. I am not sure how I could avoid this with pass-by-instance.
Since I do not need the value on caller side anymore, I switched to `swap` here. Should I rename the method to `SwapAIBattlePlayers`?

Reply:
> you are right, this approach will still create a copy. I am not sure how I could avoid this with pass-by-instance.

On the call site: `SetAIBattlePlayers(std::move(aiInfos))` and in the function move the argument. To my understanding this is the "modern C++" for sink arguments (clang-tidy even has a check for that, at least for constructors and `std::string`)

Note that the move is in this case a "cheap swap", I'm pretty sure the move-ctor/assignment of `std::vector` simply swaps 3 pointers at most. Hence the swap inside the function is not required as that's what the move will already do (at worst) so the change to an instance parameter will at worst (and likely)  just involve an addition move-construction which as mentioned is cheap enough.

> Since I do not need the value on caller side anymore

This is what `std::move(aiInfos)` at the call site expresses very well.

Having this an instance parameter allows to make a copy IF the value on the call site would be needed, hence that provides more flexibility at the cost of a (cheap) additional move.